{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Matrix Processing with Pandas\n",
    "### 2023-10-12\n",
    "\n",
    "This notebook details the step to process the matrix of allele frequencies for each variant in each sample. The matrix is generated from the output `bcftools` pipeline."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>98JKPD8_LumA_fps</th>\n",
       "      <th>ANAB5F7_Basal_fps</th>\n",
       "      <th>S6R691V_Her2_fps</th>\n",
       "      <th>PU24GB8_LumB_fps</th>\n",
       "      <th>2GAMBDQ_Norm_fps</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>191257</td>\n",
       "      <td>191268</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>818021</td>\n",
       "      <td>818032</td>\n",
       "      <td>0.02966</td>\n",
       "      <td>0.04399</td>\n",
       "      <td>0.02131</td>\n",
       "      <td>0.03470</td>\n",
       "      <td>0.03443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>818780</td>\n",
       "      <td>818791</td>\n",
       "      <td>0.02572</td>\n",
       "      <td>0.02376</td>\n",
       "      <td>0.02271</td>\n",
       "      <td>0.02814</td>\n",
       "      <td>0.03391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>960570</td>\n",
       "      <td>960581</td>\n",
       "      <td>0.28046</td>\n",
       "      <td>0.33500</td>\n",
       "      <td>0.28495</td>\n",
       "      <td>0.27618</td>\n",
       "      <td>0.22986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977015</td>\n",
       "      <td>977026</td>\n",
       "      <td>1.10047</td>\n",
       "      <td>0.50514</td>\n",
       "      <td>0.71035</td>\n",
       "      <td>1.21305</td>\n",
       "      <td>1.30155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977047</td>\n",
       "      <td>977058</td>\n",
       "      <td>0.88179</td>\n",
       "      <td>0.40974</td>\n",
       "      <td>0.53173</td>\n",
       "      <td>0.98628</td>\n",
       "      <td>1.07101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977110</td>\n",
       "      <td>977121</td>\n",
       "      <td>0.13137</td>\n",
       "      <td>0.08845</td>\n",
       "      <td>0.10541</td>\n",
       "      <td>0.16812</td>\n",
       "      <td>0.15493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977142</td>\n",
       "      <td>977153</td>\n",
       "      <td>0.14436</td>\n",
       "      <td>0.08225</td>\n",
       "      <td>0.09233</td>\n",
       "      <td>0.17809</td>\n",
       "      <td>0.14291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977174</td>\n",
       "      <td>977185</td>\n",
       "      <td>0.22616</td>\n",
       "      <td>0.10631</td>\n",
       "      <td>0.13808</td>\n",
       "      <td>0.24319</td>\n",
       "      <td>0.22306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977206</td>\n",
       "      <td>977217</td>\n",
       "      <td>0.29710</td>\n",
       "      <td>0.15487</td>\n",
       "      <td>0.21703</td>\n",
       "      <td>0.33427</td>\n",
       "      <td>0.31093</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome   Start     End  98JKPD8_LumA_fps  ANAB5F7_Basal_fps  \\\n",
       "0       chr1  191257  191268           0.00000            0.00000   \n",
       "1       chr1  818021  818032           0.02966            0.04399   \n",
       "2       chr1  818780  818791           0.02572            0.02376   \n",
       "3       chr1  960570  960581           0.28046            0.33500   \n",
       "4       chr1  977015  977026           1.10047            0.50514   \n",
       "5       chr1  977047  977058           0.88179            0.40974   \n",
       "6       chr1  977110  977121           0.13137            0.08845   \n",
       "7       chr1  977142  977153           0.14436            0.08225   \n",
       "8       chr1  977174  977185           0.22616            0.10631   \n",
       "9       chr1  977206  977217           0.29710            0.15487   \n",
       "\n",
       "   S6R691V_Her2_fps  PU24GB8_LumB_fps  2GAMBDQ_Norm_fps  \n",
       "0           0.00000           0.00000           0.00000  \n",
       "1           0.02131           0.03470           0.03443  \n",
       "2           0.02271           0.02814           0.03391  \n",
       "3           0.28495           0.27618           0.22986  \n",
       "4           0.71035           1.21305           1.30155  \n",
       "5           0.53173           0.98628           1.07101  \n",
       "6           0.10541           0.16812           0.15493  \n",
       "7           0.09233           0.17809           0.14291  \n",
       "8           0.13808           0.24319           0.22306  \n",
       "9           0.21703           0.33427           0.31093  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load an example filtered matrix of footprinting scores, which overlap the variant sites in the vcf files\n",
    "\n",
    "df_fpscore = pd.read_csv(\"/scratch/users/ntu/suffiazi/outputs/brca-vcf-filtered-fpscore-matrices/TFDP1_M08108_2.00_BRCA-subtype-vcf-filtered-matrix.txt\", sep=\"\\t\")\n",
    "\n",
    "# drop the column \"TFBS_strand\" and \"TFBS_score\"\n",
    "df_fpscore = df_fpscore.drop(columns=[\"TFBS_strand\", \"TFBS_score\"])\n",
    "# rename columns in the dataframe\n",
    "df_fpscore = df_fpscore.rename(columns={\"TFBS_chr\": \"Chromosome\", \"TFBS_start\": \"Start\", \"TFBS_end\": \"End\", \"2GAMBDQ_Normal-like_score\": \"2GAMBDQ_Norm_fps\"})\n",
    "# for all column names that end with the string 'score', replace the string with 'fps'\n",
    "df_fpscore = df_fpscore.rename(columns=lambda x: x.replace('score', 'fps') if x.endswith('score') else x)\n",
    "df_fpscore.head(n=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>818025</td>\n",
       "      <td>818025</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.812500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>818783</td>\n",
       "      <td>818783</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.033333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977023</td>\n",
       "      <td>977023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.941176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977055</td>\n",
       "      <td>977055</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977118</td>\n",
       "      <td>977118</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.852941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977150</td>\n",
       "      <td>977150</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.531250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977214</td>\n",
       "      <td>977214</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.588235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977338</td>\n",
       "      <td>977338</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977370</td>\n",
       "      <td>977370</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>986629</td>\n",
       "      <td>986629</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.156250</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome   Start     End ref_allele alt_allele        AF\n",
       "0       chr1  818025  818025          C          A  0.812500\n",
       "1       chr1  818783  818783          T          C  0.033333\n",
       "2       chr1  977023  977023          G          A  0.941176\n",
       "3       chr1  977055  977055          G          A  0.500000\n",
       "4       chr1  977118  977118          G          A  0.852941\n",
       "5       chr1  977150  977150          G          A  0.531250\n",
       "6       chr1  977214  977214          G          A  0.588235\n",
       "7       chr1  977338  977338          G          A  1.000000\n",
       "8       chr1  977370  977370          G          A  1.000000\n",
       "9       chr1  986629  986629          G          A  0.156250"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load up the vcf file with indels and multiallelic sites split into separate rows\n",
    "df_vcf_basal = pd.read_csv(\"/scratch/users/ntu/suffiazi/outputs/brca-called-variants-diffmode-AF-query/ANAB5F7/ANAB5F7_TFDP1_M08108_2.00_AF-per-site-with-indels.txt\", sep=\"\\t\")\n",
    "# rename columns in the dataframe\n",
    "df_vcf_basal = df_vcf_basal.rename(columns={\"#[1]CHROM\": \"Chromosome\", \"[2]POS\": \"Start\", \"[3]REF\": \"ref_allele\", \"[4]ALT\": \"alt_allele\", \"[5]AF\": \"AF\"})\n",
    "# add a column next to the \"start\" column called \"end\" with the same value as the \"start\" column\n",
    "df_vcf_basal.insert(2, \"End\", df_vcf_basal[\"Start\"])\n",
    "\n",
    "df_vcf_basal.head(n=10)\n",
    "# df_vcf_basal.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>818025</td>\n",
       "      <td>818025</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.900000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>818783</td>\n",
       "      <td>818783</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.125000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>960574</td>\n",
       "      <td>960574</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.041667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977023</td>\n",
       "      <td>977023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977055</td>\n",
       "      <td>977055</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977118</td>\n",
       "      <td>977118</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.909091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977150</td>\n",
       "      <td>977150</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.625000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977214</td>\n",
       "      <td>977214</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.818182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977338</td>\n",
       "      <td>977338</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977370</td>\n",
       "      <td>977370</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome   Start     End ref_allele alt_allele        AF\n",
       "0       chr1  818025  818025          C          A  0.900000\n",
       "1       chr1  818783  818783          T          C  0.125000\n",
       "2       chr1  960574  960574          C          A  0.041667\n",
       "3       chr1  977023  977023          G          A  1.000000\n",
       "4       chr1  977055  977055          G          A  1.000000\n",
       "5       chr1  977118  977118          G          A  0.909091\n",
       "6       chr1  977150  977150          G          A  0.625000\n",
       "7       chr1  977214  977214          G          A  0.818182\n",
       "8       chr1  977338  977338          G          A  1.000000\n",
       "9       chr1  977370  977370          G          A  1.000000"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load up the vcf file with indels and multiallelic sites split into separate rows\n",
    "df_vcf_lumA = pd.read_csv(\"/scratch/users/ntu/suffiazi/outputs/brca-called-variants-diffmode-AF-query/98JKPD8/98JKPD8_TFDP1_M08108_2.00_AF-per-site-with-indels.txt\", sep=\"\\t\")\n",
    "# rename columns in the dataframe\n",
    "df_vcf_lumA = df_vcf_lumA.rename(columns={\"#[1]CHROM\": \"Chromosome\", \"[2]POS\": \"Start\", \"[3]REF\": \"ref_allele\", \"[4]ALT\": \"alt_allele\", \"[5]AF\": \"AF\"})\n",
    "# add a column next to the \"start\" column called \"end\" with the same value as the \"start\" column\n",
    "df_vcf_lumA.insert(2, \"End\", df_vcf_lumA[\"Start\"])\n",
    "\n",
    "df_vcf_lumA.head(n=10)\n",
    "# df_vcf_lumA.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>191258</td>\n",
       "      <td>191258</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.026316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>818025</td>\n",
       "      <td>818025</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.833333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>818783</td>\n",
       "      <td>818783</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.055556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977023</td>\n",
       "      <td>977023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.947368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977055</td>\n",
       "      <td>977055</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.789474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977118</td>\n",
       "      <td>977118</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.944444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977150</td>\n",
       "      <td>977150</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977214</td>\n",
       "      <td>977214</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.868421</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977338</td>\n",
       "      <td>977338</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977370</td>\n",
       "      <td>977370</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome   Start     End ref_allele alt_allele        AF\n",
       "0       chr1  191258  191258          G          C  0.026316\n",
       "1       chr1  818025  818025          C          A  0.833333\n",
       "2       chr1  818783  818783          T          C  0.055556\n",
       "3       chr1  977023  977023          G          A  0.947368\n",
       "4       chr1  977055  977055          G          A  0.789474\n",
       "5       chr1  977118  977118          G          A  0.944444\n",
       "6       chr1  977150  977150          G          A  1.000000\n",
       "7       chr1  977214  977214          G          A  0.868421\n",
       "8       chr1  977338  977338          G          A  1.000000\n",
       "9       chr1  977370  977370          G          A  1.000000"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load up the vcf file with indels and multiallelic sites split into separate rows\n",
    "df_vcf_lumB = pd.read_csv(\"/scratch/users/ntu/suffiazi/outputs/brca-called-variants-diffmode-AF-query/PU24GB8/PU24GB8_TFDP1_M08108_2.00_AF-per-site-with-indels.txt\", sep=\"\\t\")\n",
    "# rename columns in the dataframe\n",
    "df_vcf_lumB = df_vcf_lumB.rename(columns={\"#[1]CHROM\": \"Chromosome\", \"[2]POS\": \"Start\", \"[3]REF\": \"ref_allele\", \"[4]ALT\": \"alt_allele\", \"[5]AF\": \"AF\"})\n",
    "# add a column next to the \"start\" column called \"end\" with the same value as the \"start\" column\n",
    "df_vcf_lumB.insert(2, \"End\", df_vcf_lumB[\"Start\"])\n",
    "\n",
    "df_vcf_lumB.head(n=10)\n",
    "# df_vcf_lumB.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>818025</td>\n",
       "      <td>818025</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.900000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>818783</td>\n",
       "      <td>818783</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.062500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977023</td>\n",
       "      <td>977023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.911765</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977055</td>\n",
       "      <td>977055</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.588235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977118</td>\n",
       "      <td>977118</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977150</td>\n",
       "      <td>977150</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977338</td>\n",
       "      <td>977338</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977370</td>\n",
       "      <td>977370</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>986629</td>\n",
       "      <td>986629</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.264706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>996350</td>\n",
       "      <td>996350</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome   Start     End ref_allele alt_allele        AF\n",
       "0       chr1  818025  818025          C          A  0.900000\n",
       "1       chr1  818783  818783          T          C  0.062500\n",
       "2       chr1  977023  977023          G          A  0.911765\n",
       "3       chr1  977055  977055          G          A  0.588235\n",
       "4       chr1  977118  977118          G          A  1.000000\n",
       "5       chr1  977150  977150          G          A  1.000000\n",
       "6       chr1  977338  977338          G          A  1.000000\n",
       "7       chr1  977370  977370          G          A  1.000000\n",
       "8       chr1  986629  986629          G          A  0.264706\n",
       "9       chr1  996350  996350          A          G  1.000000"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load up the vcf file with indels and multiallelic sites split into separate rows\n",
    "df_vcf_her2 = pd.read_csv(\"/scratch/users/ntu/suffiazi/outputs/brca-called-variants-diffmode-AF-query/S6R691V/S6R691V_TFDP1_M08108_2.00_AF-per-site-with-indels.txt\", sep=\"\\t\")\n",
    "# rename columns in the dataframe\n",
    "df_vcf_her2 = df_vcf_her2.rename(columns={\"#[1]CHROM\": \"Chromosome\", \"[2]POS\": \"Start\", \"[3]REF\": \"ref_allele\", \"[4]ALT\": \"alt_allele\", \"[5]AF\": \"AF\"})\n",
    "# add a column next to the \"start\" column called \"end\" with the same value as the \"start\" column\n",
    "df_vcf_her2.insert(2, \"End\", df_vcf_her2[\"Start\"])\n",
    "\n",
    "df_vcf_her2.head(n=10)\n",
    "# df_vcf_her2.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>818025</td>\n",
       "      <td>818025</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977023</td>\n",
       "      <td>977023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977055</td>\n",
       "      <td>977055</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977118</td>\n",
       "      <td>977118</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977150</td>\n",
       "      <td>977150</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977182</td>\n",
       "      <td>977182</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977214</td>\n",
       "      <td>977214</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977338</td>\n",
       "      <td>977338</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977370</td>\n",
       "      <td>977370</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977402</td>\n",
       "      <td>977402</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome   Start     End ref_allele alt_allele     AF\n",
       "0       chr1  818025  818025          C          A  0.875\n",
       "1       chr1  977023  977023          G          A  1.000\n",
       "2       chr1  977055  977055          G          A  0.750\n",
       "3       chr1  977118  977118          G          A  1.000\n",
       "4       chr1  977150  977150          G          A  1.000\n",
       "5       chr1  977182  977182          G          A  1.000\n",
       "6       chr1  977214  977214          G          A  1.000\n",
       "7       chr1  977338  977338          G          A  1.000\n",
       "8       chr1  977370  977370          G          A  1.000\n",
       "9       chr1  977402  977402          G          A  1.000"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load up the vcf file with indels and multiallelic sites split into separate rows\n",
    "df_vcf_norm = pd.read_csv(\"/scratch/users/ntu/suffiazi/outputs/brca-called-variants-diffmode-AF-query/2GAMBDQ/2GAMBDQ_TFDP1_M08108_2.00_AF-per-site-with-indels.txt\", sep=\"\\t\")\n",
    "# rename columns in the dataframe\n",
    "df_vcf_norm = df_vcf_norm.rename(columns={\"#[1]CHROM\": \"Chromosome\", \"[2]POS\": \"Start\", \"[3]REF\": \"ref_allele\", \"[4]ALT\": \"alt_allele\", \"[5]AF\": \"AF\"})\n",
    "# add a column next to the \"start\" column called \"end\" with the same value as the \"start\" column\n",
    "df_vcf_norm.insert(2, \"End\", df_vcf_norm[\"Start\"])\n",
    "\n",
    "df_vcf_norm.head(n=10)\n",
    "# df_vcf_norm.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>ref_allele</th>\n",
       "      <th>alt_allele</th>\n",
       "      <th>AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>818025</td>\n",
       "      <td>818025</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>0.812500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>818783</td>\n",
       "      <td>818783</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.033333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977023</td>\n",
       "      <td>977023</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.941176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977055</td>\n",
       "      <td>977055</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.500000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>977118</td>\n",
       "      <td>977118</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.852941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5979</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153494830</td>\n",
       "      <td>153494830</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.352941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5980</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153495413</td>\n",
       "      <td>153495413</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.882353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5981</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154444635</td>\n",
       "      <td>154444635</td>\n",
       "      <td>GGGACGG</td>\n",
       "      <td>GGGACGGACGG</td>\n",
       "      <td>0.058824</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5982</th>\n",
       "      <td>chrY</td>\n",
       "      <td>11108021</td>\n",
       "      <td>11108021</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.382353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5983</th>\n",
       "      <td>chrY</td>\n",
       "      <td>11108024</td>\n",
       "      <td>11108024</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.294118</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5984 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Chromosome      Start        End ref_allele   alt_allele        AF\n",
       "0          chr1     818025     818025          C            A  0.812500\n",
       "1          chr1     818783     818783          T            C  0.033333\n",
       "2          chr1     977023     977023          G            A  0.941176\n",
       "3          chr1     977055     977055          G            A  0.500000\n",
       "4          chr1     977118     977118          G            A  0.852941\n",
       "...         ...        ...        ...        ...          ...       ...\n",
       "5979       chrX  153494830  153494830          G            A  0.352941\n",
       "5980       chrX  153495413  153495413          G            A  0.882353\n",
       "5981       chrX  154444635  154444635    GGGACGG  GGGACGGACGG  0.058824\n",
       "5982       chrY   11108021   11108021          C            T  0.382353\n",
       "5983       chrY   11108024   11108024          C            T  0.294118\n",
       "\n",
       "[5984 rows x 6 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_vcf_basal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------+-----------+-----------+--------------------+-------+\n",
      "| Chromosome   | Start     | End       | 98JKPD8_LumA_fps   | +4    |\n",
      "| (category)   | (int64)   | (int64)   | (float64)          | ...   |\n",
      "|--------------+-----------+-----------+--------------------+-------|\n",
      "| chr1         | 191257    | 191268    | 0.0                | ...   |\n",
      "| chr1         | 818021    | 818032    | 0.02966            | ...   |\n",
      "| chr1         | 818780    | 818791    | 0.02572            | ...   |\n",
      "| chr1         | 960570    | 960581    | 0.28046            | ...   |\n",
      "| ...          | ...       | ...       | ...                | ...   |\n",
      "| chrX         | 155641610 | 155641621 | 0.01249            | ...   |\n",
      "| chrX         | 155010397 | 155010408 | 0.01177            | ...   |\n",
      "| chrX         | 154444629 | 154444640 | 0.049              | ...   |\n",
      "| chrX         | 153934861 | 153934872 | 0.1967             | ...   |\n",
      "| chrY         | 11108019  | 11108030  | 0.0                | ...   |\n",
      "+--------------+-----------+-----------+--------------------+-------+\n",
      "Unstranded PyRanges object has 9,560 rows and 8 columns from 24 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome.\n",
      "4 hidden columns: ANAB5F7_Basal_fps, S6R691V_Her2_fps, PU24GB8_LumB_fps, ... (+ 1 more.)\n"
     ]
    }
   ],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "gr_fpscore = pr.PyRanges(df_fpscore)\n",
    "print(gr_fpscore)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+--------------+-----------+-----------+--------------+-------+\n",
      "| Chromosome   | Start     | End       | ref_allele   | +2    |\n",
      "| (category)   | (int64)   | (int64)   | (object)     | ...   |\n",
      "|--------------+-----------+-----------+--------------+-------|\n",
      "| chr1         | 818025    | 818025    | C            | ...   |\n",
      "| chr1         | 818783    | 818783    | T            | ...   |\n",
      "| chr1         | 977023    | 977023    | G            | ...   |\n",
      "| chr1         | 977055    | 977055    | G            | ...   |\n",
      "| ...          | ...       | ...       | ...          | ...   |\n",
      "| chrX         | 154444635 | 154444635 | GGGACGG      | ...   |\n",
      "| chrX         | 153495413 | 153495413 | G            | ...   |\n",
      "| chrX         | 153494830 | 153494830 | G            | ...   |\n",
      "| chrX         | 150364545 | 150364545 | C            | ...   |\n",
      "| chrY         | 11108024  | 11108024  | C            | ...   |\n",
      "| chrY         | 11108021  | 11108021  | C            | ...   |\n",
      "+--------------+-----------+-----------+--------------+-------+\n",
      "Unstranded PyRanges object has 5,984 rows and 6 columns from 24 chromosomes.\n",
      "For printing, the PyRanges was sorted on Chromosome.\n",
      "2 hidden columns: alt_allele, AF\n"
     ]
    }
   ],
   "source": [
    "gr_vcfBasal = pr.PyRanges(df_vcf_basal)\n",
    "print(gr_vcfBasal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import bioframe as bf\n",
    "\n",
    "# # overlap fpscore combined matrix with Basal vcf data\n",
    "# overlap_df = bf.overlap(df_fpscore, df_vcf_basal, suffixes=[\"\",\"_ANAB5F7_Basal_varsite\"], how='left')\n",
    "\n",
    "# # drop these columns\n",
    "# drop_patterns = [\"ref_allele_\", \"alt_allele_\", \"chrom_\", \"end_\"]\n",
    "# for pat in drop_patterns:\n",
    "#     overlap_df = overlap_df.drop(columns=overlap_df.filter(like=pat, axis=1).columns)\n",
    "# # rename these columns\n",
    "# overlap_df = overlap_df.rename(columns=lambda x: x.replace('start_', 'pos_') if x.startswith('start_') else x)\n",
    "\n",
    "# # overlap the resulting dataframe again with lumA vcf data\n",
    "# overlap_df = bf.overlap(overlap_df, df_vcf_lumA, suffixes=[\"\",\"_98JKPD8_LumA_varsite\"], how='left')\n",
    "\n",
    "# # drop these columns\n",
    "# for pat in drop_patterns:\n",
    "#     overlap_df = overlap_df.drop(columns=overlap_df.filter(like=pat, axis=1).columns)\n",
    "# # rename these columns\n",
    "# overlap_df = overlap_df.rename(columns=lambda x: x.replace('start_', 'pos_') if x.startswith('start_') else x)\n",
    "\n",
    "# # overlap again with lumB vcf data\n",
    "# overlap_df = bf.overlap(overlap_df, df_vcf_lumB, suffixes=[\"\",\"_PU24GB8_LumB_varsite\"], how='left')\n",
    "\n",
    "# # drop these columns\n",
    "# for pat in drop_patterns:\n",
    "#     overlap_df = overlap_df.drop(columns=overlap_df.filter(like=pat, axis=1).columns)\n",
    "# # rename these columns\n",
    "# overlap_df = overlap_df.rename(columns=lambda x: x.replace('start_', 'pos_') if x.startswith('start_') else x)\n",
    "\n",
    "# # overlap again with her2 vcf data\n",
    "# overlap_df = bf.overlap(overlap_df, df_vcf_her2, suffixes=[\"\",\"_S6R691V_Her2_varsite\"], how='left')\n",
    "\n",
    "# # drop these columns\n",
    "# for pat in drop_patterns:\n",
    "#     overlap_df = overlap_df.drop(columns=overlap_df.filter(like=pat, axis=1).columns)\n",
    "# # rename these columns\n",
    "# overlap_df = overlap_df.rename(columns=lambda x: x.replace('start_', 'pos_') if x.startswith('start_') else x)\n",
    "\n",
    "# # overlap again with normal-like vcf data\n",
    "# overlap_df = bf.overlap(overlap_df, df_vcf_norm, suffixes=[\"\",\"_2GAMBDQ_Norm_varsite\"], how='left')\n",
    "\n",
    "# # drop these columns\n",
    "# for pat in drop_patterns:\n",
    "#     overlap_df = overlap_df.drop(columns=overlap_df.filter(like=pat, axis=1).columns)\n",
    "# # rename these columns\n",
    "# overlap_df = overlap_df.rename(columns=lambda x: x.replace('start_', 'pos_') if x.startswith('start_') else x)\n",
    "\n",
    "# # save the dataframe to a file\n",
    "# overlap_df.to_csv('/home/users/ntu/suffiazi/scripts/gatk-workflow-scripts/polars_processing/combined_fps-variant-freq.tsv', sep='\\t', index=False)\n",
    "\n",
    "# overlap_df.head(n=15)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# show only row number 100 and 101\n",
    "# overlap_df.iloc[[99, 100]]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gatk-snv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
