#!/usr/bin/env bash
# shellcheck disable=SC1091

#PBS -l select=1:ncpus=2:mem=100GB
#PBS -l walltime=24:00:00
#PBS -j oe
#PBS -P personal
#PBS -q normal

# load conda environment
module load miniconda3/py38_4.8.3

conda activate /home/users/ntu/suffiazi/apps/mambaforge/envs/gatk-snv

# set up variables
tfbs_loc=$TF_LIST
bams=$BAM_INP
outfile=$OUT_DIR
data_id=$ID

# read in TF motif list
readarray -t tfbs_list < "$tfbs_loc"
echo "Number of TF motif bed files:" "${#tfbs_list[@]}"

# loop through each TF motif
for bedfile in "${tfbs_list[@]}"; do
    echo "TF motif of interest: $bedfile"
    # get motif ID
    motif_id="${bedfile%_tfbs_merged_matrix-brca_brca.bed}"
    echo "Motif ID: $motif_id"

    # check if output vcf file already exists
    if [ -f "${outfile}"/"${data_id}"_"${motif_id}"_qualgt10.var.flt.vcf ]; then
        echo "Output vcf file already exists. Skipping..."
        continue
    else
        echo "Output vcf file does not exist. Proceeding..."
        # get bedfile path
        bed_path=$(find /scratch/users/ntu/suffiazi/inputs/brca_tfbs_bedfiles -name "${bedfile}" -type f)
        if [ -n "$bed_path" ]; then
            echo "Found bed file. Proceeding..."
            echo "Bed file path: $bed_path"
            # call and filter variants from bam files of BRCA datasets
            if bcftools mpileup -Ou -f /scratch/users/ntu/suffiazi/inputs/references/gatk4/GRCh38_no_alt_analysis_set_GCA_000001405.15.fasta -T "${bed_path}" -b "${bams}" | bcftools call -Ou -mv | bcftools filter -i'QUAL>10' > "${outfile}"/"${data_id}"_"${motif_id}"_qualgt10.var.flt.vcf; then
                echo "Variant calling and filtering completed."
            else
                echo "Variant calling and filtering failed."
                exit 1
            fi
        else
            echo "Bed file ${bedfile} not found. Check again. Exiting..."
            exit 1
        fi
    fi   
done
