#!/usr/bin/env bash
# shellcheck disable=SC1091

#PBS -l select=1:ncpus=12:mem=120GB
#PBS -l walltime=8:00:00
#PBS -j oe
#PBS -P 12003580
#PBS -q normal

set -o pipefail

# load conda environment
module load singularity
singExec="singularity exec --bind /home/users/ntu/suffiazi/scratch:/mnt /home/users/ntu/suffiazi/apps/sifs/gatk.sif"


# set up variable
file_prefix=$PREFIX

# run gatk SamToFastq
$singExec gatk --java-options "-Xmx80G" SamToFastq \
        I="${file_prefix}"_markillum-adapt.bam \
        FASTQ=/dev/stdout \
        CLIPPING_ATTRIBUTE=XT \
        CLIPPING_ACTION=2 \
        INTERLEAVE=true \
        NON_PF=true \
        TMPDIR=/mnt/outputs/tmp | \

bwa mem -M -t 7 -p /path/Homosapiens_assembly19.fasta /dev/stdin | \

gatk --java-options "-Xmx80G" MergeBamAlignment \
        ALIGNED_BAM=/dev/stdin \
        UNMAPPED_BAM="${file_prefix}"_revertsam.bam \
        OUTPUT="${file_prefix}"_pipe-merged.bam \
        R=/path/Homo_sapiens_assembly19.fasta \
        CREATE_INDEX=true \
        ADD_MATE_CIGAR=true \
        CLIP_ADAPTERS=false \
        CLIP_OVERLAPPING_READS=true \
        INCLUDE_SECONDARY_ALIGNMENTS=true \
        MAX_INSERTIONS_OR_DELETIONS=-1 \
        PRIMARY_ALIGNMENT_STRATEGY=MostDistant \
        ATTRIBUTES_TO_RETAIN=XS \
        TMP_DIR=/mnt/outputs/tmp